triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -264,44 +267,68 @@ class Runner(
-    // We use the queue to post failures that occur directly on command submission as opposed to
-    // appearing asynchronously on the completion stream
-    val (completionQueue, completionQueueSource) =
+
+    // A queue for command submission failures together with a source
+    // from which messages posted to the queue can be consumed.
+    val (completionQueue, completionQueueSource): (
+        SourceQueueWithComplete[Completion],
+        Source[Completion, NotUsed]) =
       Source.queue[Completion](10, OverflowStrategy.backpressure).preMaterialize()
-    val transactionSource =
+    // This function, given a command ID and a runtime exception,
+    // posts to the completion queue.
+    def postSubmitFailure(commandId: String, s: StatusRuntimeException) = {
+      // 'Completion' is a type generated by protoc. It wraps a grpc
+      // status code and represents the status of a committed command
+      // on the ledger. Here we produce a value of this type corresponding
+      // to a failed command submission.
+      val _ = completionQueue.offer(
+        Completion(
+          commandId,
+          Some(Status(s.getStatus().getCode().value(), s.getStatus().getDescription()))))
+    }
+
+    // The transaction source (ledger).
+    val transactionSource: Source[TriggerMsg, NotUsed] =
       client.transactionClient
         .getTransactions(offset, None, filter, verbose = true)
-        .map[TriggerMsg](TransactionMsg)
-    val completionSource =
+        .map(TransactionMsg)
+
+    // Command completion source (ledger + failures).
+    val completionSource: Source[TriggerMsg, NotUsed] =
       client.commandClient
         .completionSource(List(party), offset)
         .mapConcat({
           case CheckpointElement(_) => List()
           case CompletionElement(c) => List(c)
-        })
-        .merge(completionQueueSource)
-        .map[TriggerMsg](CompletionMsg)
-    val source = heartbeat match {
+        }) // Successful completions.
+        .merge(completionQueueSource) // Failed completions.
+        .map(CompletionMsg)
+
+    // Hearbeats source (we produce these repetitvely on a timer with
+    // the given delay interval).
+    val heartbeatSource: Source[TriggerMsg, NotUsed] = heartbeat match {
       case Some(interval) =>
-        transactionSource
-          .merge(completionSource)
-          .merge(Source.tick[TriggerMsg](interval, interval, HeartbeatMsg()))
-      case None => transactionSource.merge(completionSource)
-    }
-    def postSubmitFailure(commandId: String, s: StatusRuntimeException) = {
-      val _ = completionQueue.offer(
-        Completion(
-          commandId,
-          Some(Status(s.getStatus().getCode().value(), s.getStatus().getDescription()))))
+        val source: Source[TriggerMsg, Cancellable] =
+          Source.tick[TriggerMsg](interval, interval, HeartbeatMsg())
+        source.asInstanceOf[Source[TriggerMsg, NotUsed]] // Upcast. Won't fail.
```

Comment:
I don't understand that. I claim the code the code is just the original program with explicit type annotations.

Without annotations we have:
```
    val heartbeatSource = heartbeat match {
      case Some(interval) =>
          Source.tick[TriggerMsg](interval, interval, HeartbeatMsg())
      case None => Source.empty
    }
```

Now If I annotate `heartbeatSource` as `heartbeatSource:[TriggerMsg, NotUsed]` I get a type error:
```
error: type mismatch;
 found   : akka.stream.scaladsl.Source[com.daml.lf.engine.trigger.TriggerMsg,akka.actor.Cancellable]
 required: akka.stream.scaladsl.Source[com.daml.lf.engine.trigger.TriggerMsg,akka.NotUsed]
          Source.tick[TriggerMsg](interval, interval, HeartbeatMsg())
```
So I followed the types and we are now looking at the result.

My interpretation is without annotation, an implicit upcast is performed. When I annotate, the implicit cast needs to be made explicit.

Reply:
>     val heartbeatSource = heartbeat match {
>       case Some(interval) =>
>           Source.tick[TriggerMsg](interval, interval, HeartbeatMsg())
>       case None => Source.empty
>     }

This is not what we had before. Before we never got into a situation where we had to assign the same type to `Source.empty` and `Source.tick…` because we directly merged in the two case branches.

Reply:
That's a good point.

Reply:
The fact remains though. 
```
val heartbeatSource = heartbeat match {
  case Some(interval) =>
      Source.tick[TriggerMsg](interval, interval, HeartbeatMsg())
  case None => Source.empty
}
```
This types and one assumes it types to `Source[TriggerMsg, NotUsed]`. If I annotate as such though then it no longer types without the cast.

Reply:
I'll rephrase it so the cast is gone. I'd like to understand it though in the end. Perhaps I've incorrectly calculated the type of `heartbeat`?

Reply:
The inferred type is `Source[TriggerMsg, Object]` not `Source[TriggerMsg, NotUsed]` which is an actual upcast. Again, `NotUsed` is not a superclass of `Cancellable` and scala will not do random casts during type inference. You could annotate with `Source[TriggerMsg, Object]` but I think `Source[TriggerMsg, NotUsed]` + `mapMaterializedValue` is a better solution here.

Reply:
Ok thanks. Yes, I was banking on the assumption that `Cancellable` was a sub-type of `NotUsed`.